capsules/extra/src/sht4x.rs

```diff
@@ -0,0 +1,246 @@
+
+                // buffer[0] = ((Registers::MEASHIGHREP as u16) >> 8) as u8;
+                // buffer[1] = ((Registers::MEASHIGHREP as u16) & 0xff) as u8;
+                buffer[0] = Registers::MEASHIGHREP as u8;
+
+                // TODO verify errors
+                let _ = self.i2c.write(buffer, 1);
+
+                Ok(())
+            },
+        )
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: i2c::I2CDevice> time::AlarmClient for SHT4x<'a, A, I> {
+    fn alarm(&self) {
+        let state = self.state.get();
+        match state {
+            State::Read => {
+                self.state.set(State::ReadData);
+                self.buffer.take().map_or_else(
+                    || panic!("SHT4x No buffer available!"),
+                    |buffer| {
+                        let _res = self.i2c.read(buffer, 6);
+                    },
+                );
+            }
+            _ => {
+                // This should never happen
+                panic!("SHT4x Invalid alarm!");
+            }
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: i2c::I2CDevice> i2c::I2CClient for SHT4x<'a, A, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        match status {
+            Ok(()) => {
+                let state = self.state.get();
+
+                match state {
+                    State::ReadData => {
+                        if self.read_temp.get() == true {
+                            self.read_temp.set(false);
+                            if crc8(&buffer[0..2]) == buffer[2] {
+                                let mut stemp = buffer[0] as u32;
+                                stemp = stemp << 8;
+                                stemp = stemp | buffer[1] as u32;
+                                let stemp = ((4375 * stemp) >> 14) as i32 - 4500;
+                                self.temperature_client.map(|cb| cb.callback(Ok(stemp)));
+                            } else {
+                                self.temperature_client
+                                    .map(|cb| cb.callback(Err(ErrorCode::FAIL)));
+                            }
+                        }
+                        if self.read_hum.get() == true {
+                            self.read_hum.set(false);
+                            if crc8(&buffer[3..5]) == buffer[5] {
+                                let mut shum = buffer[3] as u32;
+                                shum = shum << 8;
+                                shum = shum | buffer[4] as u32;
+                                shum = (625 * shum) >> 12;
+                                self.humidity_client.map(|cb| cb.callback(shum as usize));
+                            } else {
+                                self.humidity_client.map(|cb| cb.callback(usize::MAX));
+                            }
+                        }
+                        self.buffer.replace(buffer);
+                        self.state.set(State::Idle);
```

Comment:
This is subtle, but these two lines need to happen before any client callbacks are called.

What can happen right now is:

1. This capsule does a temp/humi reading and issues a callback.
2. The client capsule gets the reading and decides it wants another reading.
3. The client capsule then calls TemperatureDriver::get().
4. This capsules returns busy, even though it isn't really busy, because the state hasn't been reset yet.

By replacing the buffer and setting the state to idle _before_ issuing the callback this issue can be avoided.